### PR TITLE
LLVM flow for V programs

### LINKED ISSUE #1: Compatibility with RVV 0.10
Once the RISC-V GCC toolchain and the Spike simulator have been updated to version 0.10, the same can be done with Ara. 

In terms of microarchitectural changes, RVV 0.10 deprecates the `SLEN` parameter, the striping distance being set as being always equal to the vector length `VLEN`. Current Ara already takes this into account, since it considers that `SLEN = VLEN`.

Other RVV-upgrade changes include the implementation of extra floating-point instructions, and encoding changes with a few vector CSRs.

### PR DESCRIPTION
**Pull request by mp-17**
# Update Compile flow to LLVM, and Ara to RVV 0.10

This PR is quite huge since I modified both the software and the hardware to make the CI pipeline pass.

The main motivation to replace GCC with LLVM is summarized [here](https://github.com/riscv/riscv-gnu-toolchain/issues/871): 
> RVV work in general is focused on LLVM at the moment. [jim-wilson]

LLVM seems to work quite decently, and it is compatible with the [official RISC-V Vector intrinsics](https://github.com/riscv/rvv-intrinsic-doc).
 
## Software:
- The reference toolchain is now LLVM, and the patches to the toolchain and SPIKE are no more necessary.
- LLVM needs a separate installation of the toolchain and the related libraries.
- [newlib](https://sourceware.org/newlib/) (`libc`, `libm`, ...) is taken from its official source, and the low-level [LLVM compiler-rt](https://github.com/llvm/llvm-project/tree/main/compiler-rt) (replacement for `libgcc`) is taken from the official LLVM repository.
- GCC is still used to compile the scalar `riscv-tests`, as some parts of the tests are written using GCC-specific dialect.
- SPIKE is updated to support RVV 0.10.
- The CI now supports LLVM and the necessary libraries.
- The Makefiles supports now LLVM and its options. It's not possible to link files compiled with the two different toolchains (GCC and LLVM). Therefore, the common/runtime source files must be re-compiled when the toolchain is changed. This happens when compiling riscv-tests since the scalar tests are compiled using GCC, while the vector ones are compiled with LLVM.
- To be compatible with the testbench we use, every .elf section must be aligned to the AXI data bus width in the Linker Script. A special script is always called by the Makefile to modify this alignment, which depends on the configuration parameter `NrLanes`.
- The benchmarks were updated to RVV 0.10 (small changes).
- Minimal compatibility modifications were done to riscv-tests, only to the common files used by the vector tests, which are compiled with LLVM.
- I added new tests for new RVV 0.10 instructions, strided memory operations, misaligned (wrt AXI data bus width) load/stores

## Hardware:
- Update new vtype encoding
- Add support for `vsetivli` instruction
- Add support for strided memory operations
- Add support for stores misaligned wrt AXI data bus width
- Add support for `vmv<nr>r.v` instruction
- Memory operations with vl == 0 are now processed correctly

Fixes #1.

## Checklist

- [x] Automated tests pass
- [x] Changelog updated
- [x] Code style guideline is observed

Please check our [contributing guidelines](CONTRIBUTING.md) before opening a Pull Request.

### CODE REVIEW

File: apps/riscv-tests/isa/Makefile [hunk 1]
```diff
@@ -48,13 +57,13 @@ vpath %.S $(src_dir)
 	$(RISCV_OBJDUMP) $< > $@
 
 %.out: %
-	$(RISCV_SIM) --isa=rv64gcv --varch="vlen:2048,elen:64,slen:2048" $< 2> $@
+	$(RISCV_SIM) --isa=rv64gcv --varch="vlen:4096,elen:64,slen:4096" $< 2> $@
```
Comment (suehtamacv) [inline apps/riscv-tests/isa/Makefile:60]: Any specific reason behind this change? (Should we adapt it to Ara's?)

File: apps/riscv-tests/isa/macros/vector/float_macros.h [hunk 2]
```diff
@@ -67,7 +67,7 @@ typedef union double_hex {
     uint64_t      ff;                                                                                  \
     asm volatile ("frflags %0" : "=r" (ff));                                                           \
     if (ff != gold_ff) {                                                                               \
-      printf("fflags check FAILED. Current fflags is 0x%02x, while expecting 0x%02x.\n", ff, gold_ff); \
+      printf("fflags check FAILED. Current fflags is 0x%02lx, while expecting 0x%02lx.\n", ff, gold_ff); \
```
Comment (suehtamacv) [inline apps/riscv-tests/isa/macros/vector/float_macros.h:70]: Check the vertical alignment.

File: apps/riscv-tests/isa/macros/vector/vector_macros.h [hunk 3]
```diff
@@ -27,6 +27,7 @@
 #define enable_fp()
 #endif
 
+
```
Comment (suehtamacv) [inline apps/riscv-tests/isa/macros/vector/vector_macros.h:30]: Unnecessary whitespace.

File: apps/riscv-tests/isa/macros/vector/vector_macros.h [hunk 4]
```diff
@@ -73,7 +88,7 @@ int test_case;
     num_failed++;                                                       \
     return;                                                             \
   }                                                                     \
-  printf("PASSED.\n", casenum);                                         \
+  printf("PASSED.\n");                                         \
```
Comment (suehtamacv) [inline apps/riscv-tests/isa/macros/vector/vector_macros.h:91]: Check the vertical alignment. Should we add the `casenum` in the `printf`  text?

File: apps/riscv-tests/isa/macros/vector/vector_macros.h [hunk 5]
```diff
@@ -82,7 +97,7 @@ int test_case;
     num_failed++;                                                       \
     return;                                                             \
   }                                                                     \
-  printf("PASSED.\n", casenum);
+  printf("PASSED.\n");
```
Comment (suehtamacv) [inline apps/riscv-tests/isa/macros/vector/vector_macros.h:100]: Check the vertical alignment.

File: apps/riscv-tests/isa/macros/vector/vector_macros.h [hunk 6]
```diff
@@ -96,18 +111,18 @@ int test_case;
       return;                                                                         \
     }                                                                                 \
   }                                                                                   \
-  printf("PASSED.\n", casenum);
+  printf("PASSED.\n");
```
Comment (suehtamacv) [inline apps/riscv-tests/isa/macros/vector/vector_macros.h:114]: Check the vertical alignment.

File: apps/riscv-tests/isa/rv64uv/vmerge.c [hunk 7]
```diff
@@ -62,7 +62,7 @@ void TEST_CASE2() {
   VLOAD_64(v1, 1, 2, 3, 4, 5, 6, 7, 8, 1, 2, 3, 4, 5, 6, 7, 8);
   VLOAD_8(v0, 0xAA, 0x55);
   asm volatile ("vmerge.vxm v3, v1, %[A], v0" :: [A] "r" (scalar));
-  VCMP_U64(8, v3, 1, 0xffffffffdeadbeef, 3, 0xffffffffdeadbeef, 5, 0xffffffffdeadbeef, 7, 0xffffffffdeadbeef, 0xffffffffdeadbeef, 2, 0xffffffffdeadbeef, 4, 0xffffffffdeadbeef, 6, 0xffffffffdeadbeef, 8);
```
Comment (suehtamacv) [inline apps/riscv-tests/isa/rv64uv/vmerge.c:65]: Such a weird bug.. Is this GCC's fault?

File: hardware/src/vlsu/addrgen.sv [hunk 8]
```diff
@@ -195,6 +195,37 @@ module addrgen import ara_pkg::*; import rvv_pkg::*; #(
     end
   end
 
+  /////////////////////////////////////
+  //  Support for misaligned stores  //
+  /////////////////////////////////////
+
+  // Narrower AXI Data Byte-Width used for misaligned stores
+  logic [$clog2(AxiDataWidth/8)-1:0] narrow_AxiDataBWidth;
```
Comment (suehtamacv) [inline hardware/src/vlsu/addrgen.sv:203]: Please adapt the name of the signal to the coding guidelines, by using `snake_case`.

File: hardware/src/vlsu/addrgen.sv [hunk 9]
```diff
@@ -207,6 +238,8 @@ module addrgen import ara_pkg::*; import rvv_pkg::*; #(
   axi_addr_t aligned_start_addr_d, aligned_start_addr_q;
   axi_addr_t aligned_end_addr_d, aligned_end_addr_q;
 
+  logic [$clog2(AxiDataWidth/8):0] eff_AxiDataBWidth_d, eff_AxiDataBWidth_q;
```
Comment (suehtamacv) [inline hardware/src/vlsu/addrgen.sv:241]: Please adapt the name of the signal to the coding guidelines, by using `snake_case`.

**Review by suehtamacv — CHANGES_REQUESTED** [review]
LGTM, Perotti! I have a few minor comments, but you can merge this back into `main` once you adapt the documentation.

File: apps/riscv-tests/isa/Makefile (hunk 1 above)
Comment (mp-17) [reply apps/riscv-tests/isa/Makefile:60]: It was to adapt it to Ara (always useful for debug purposes) :)

File: hardware/src/vlsu/addrgen.sv [hunk 10]
```diff
@@ -234,15 +269,23 @@ module addrgen import ara_pkg::*; import rvv_pkg::*; #(
...
+          // the store unit does not support misalignments between the Axi Bus and the
+          // data received from the lanes
+          eff_axi_data_bwidth_d =
+            ((axi_addrgen_d.addr[$clog2(AxiDataWidth/8)-1:0] != '0) && !axi_addrgen_d.is_load) ?
+            {1'b0, narrow_axi_data_bwidth} :
+            AxiDataWidth/8;
+
+          // The start address is found by aligning the original request address by the width of
+          // the memory interface.
+          aligned_start_addr_d = aligned_addr(axi_addrgen_d.addr, $clog2(eff_axi_data_bwidth_d));
```
Comment (suehtamacv) [inline hardware/src/vlsu/addrgen.sv:282]: Can you synthesize this: `$clog2(eff_axi_data_bwidth_d)` ?

**Review by suehtamacv — COMMENTED** [review]
I have a single comment about a construct that I am not sure whether you can synthesize. Besides that, it looks good to me.

File: hardware/src/vlsu/addrgen.sv (hunk 10 above)
Comment (mp-17) [reply hardware/src/vlsu/addrgen.sv]: No, thanks for noticing! Now the code is simplified, and I removed the `$clog2` of non-constant signals.

**Review by suehtamacv — APPROVED** [review]
LGTM, thanks a ton for your great work! :tada:

**Comment by mp-17** [comment]
I have only updated the Readme ;)
